metaflow/plugins/kfp/kfp.py

```diff
@@ -406,74 +402,70 @@ def create_kfp_pipeline_from_flow_graph(self):
-                        resource_requirements["memory"]
-                    )
-                if "memory_limit" in resource_requirements:
-                    container_op.container.set_memory_limit(
-                        resource_requirements["memory_limit"]
-                    )
-                if "cpu" in resource_requirements:
-                    container_op.container.set_cpu_request(resource_requirements["cpu"])
-                if "cpu_limit" in resource_requirements:
-                    container_op.container.set_cpu_limit(
-                        resource_requirements["cpu_limit"]
-                    )
-                if "gpu" in resource_requirements:
-                    # TODO(yunw)(AIP-2048): Support mixture of GPU from different vendors.
-                    container_op.container.set_gpu_limit(
-                        resource_requirements["gpu"],
-                        vendor=resource_requirements["gpu_vendor"],
-                    )
-
-            def build_kfp_dag(node: DAGNode, context: str, index=None):
-                kfp_component = step_to_kfp_component_map[node.name]
-                visited[node.name] = step_op(
+                op = visited[node.name] = step_op(
                     datastore_root,
-                    kfp_component.step_command,
-                    kfp_run_id,
-                    context,
-                    index=index,
+                    step_to_kfp_component_map[node.name].cmd_template,
+                    kfp_run_id=f"kfp-{dsl.RUN_ID_PLACEHOLDER}",
+                    passed_in_split_indexes=passed_in_split_indexes,
                 ).set_display_name(node.name)
 
-                set_resource_requirements(
-                    visited[node.name],
-                    KubeflowPipelines._get_resource_requirements(node),
+                KubeflowPipelines._set_container_settings(
+                    op, step_to_kfp_component_map[node.name]
                 )
 
                 if node.type == "foreach":
-                    with kfp.dsl.ParallelFor(
-                        visited[node.name].outputs["split_indexes"]
-                    ) as index:
-                        for step in node.out_funcs:
-                            build_kfp_dag(
-                                self.graph[step],
-                                visited[node.name].outputs["task_out_dict"],
-                                index,
-                            )
-                    # TODO: the join
-                    # visited[node.name].outputs["task_out_dict"]
+                    # Please see nested_parallelfor.ipynb for how this works
+                    with kfp.dsl.ParallelFor(op.output) as index:
+                        # build_kfp_dag() will halt when
+                        # a foreach join is reached.
+                        build_kfp_dag(
+                            self.graph[node.out_funcs[0]], passed_in_split_indexes=index
```

Comment:
Is there ever a scenario where this `node` would have > 1 `out_funcs`? If so what would happen in that case?

Reply:
added comment